sw/device/lib/base/memory_hardened.c

```diff
@@ -0,0 +1,703 @@
+#include "sw/device/lib/base/memory.h"
+
+// Convinience wrapper to return kHardenedBoolTrue if val == 0
+static hardened_bool_t true_if_zero(uintptr_t val) {
+  return (hardened_bool_t)hardened_select_if_zero(val, kHardenedBoolFalse,
+                                                  kHardenedBoolTrue);
+}
+
+// Define machine word alignment mask.
+#define WORD_MASK (sizeof(uintptr_t) - 1)
+
+static size_t min(size_t a, size_t b) { return (a < b) ? a : b; }
+
+// The 32 bit LFSR whose maximum length feedback polynomial is represented
+// as X^32 + X^22 + X^2 + X^1 + 1 will produce 2^32-1 PN sequence.
+// This LFSR can be initialized with 0,  but can't be initialized with
+// 0xFFFFFFFF. This is handy to avoid non-zero initial values.
+static uint32_t next_fast_random(uint32_t seed) {
+  // 12                        22          32
+  // 1100 0000 0000 0000 0000 0100 0000 0001 = 0xC0000401
+  uint32_t mask = (-((int32_t)seed >= 0)) & 0xC0000401;
+  return (seed << 1) ^ mask;
+}
+
+// Division by multiplication on mod 2^32 inverse.
+// Result can be larger than interfer division.
+static uint32_t div_by_mul(uint32_t p, uint32_t inv) {
+  if (p == 0)
+    return p;
+
+  // Compiles into single 'mulhu' instruction, don't use libgcc
+  uint64_t prod = (uint64_t)p * (uint64_t)inv;
+  return prod >> 32;
+}
+
+static uintptr_t read_uintptr(uintptr_t addr) {
+  // Both GCC and Clang optimize the code below into a single word-load on most
+  // platforms. It is necessary and sufficient to indicate to the compiler that
+  // the pointer points to four bytes of four-byte-aligned memory.
+  void *ptr = __builtin_assume_aligned((const void *)addr, alignof(uintptr_t));
+  uintptr_t val;
+  __builtin_memcpy(&val, ptr, sizeof(val));
+  return val;
+}
+
+static inline void write_uintptr(uintptr_t value, uintptr_t addr) {
+  // Both GCC and Clang optimize the code below into a single word-store on most
+  // platforms.
+  void *ptr = __builtin_assume_aligned((void *)addr, alignof(uintptr_t));
+  __builtin_memcpy(ptr, &value, sizeof(value));
+}
+
+hardened_bool_t memcpy_checked(void *restrict dest, const void *restrict src,
+                               size_t len) {
+  uintptr_t dest_addr = (uintptr_t)dest;
+  uintptr_t src_addr = (uintptr_t)src;
+  uintptr_t tail = dest_addr + len;
+
+  // Make a copy of tail which avoids compiler optimization.
+  volatile uintptr_t tail_copy = tail;
+
+  // Set 'body' to the last word boundary.
+  uintptr_t body = tail & ~WORD_MASK;
+
+  // End of head is the tail if data is not aligned.
+  uintptr_t head = tail;
+
+  // Compute starting address if src and dest can be aligned.
+  if ((((dest_addr ^ src_addr) & WORD_MASK) == 0) &&
+      (len >= sizeof(uintptr_t))) {
+    /* Set 'head' to the first word boundary */
+    head = ((dest_addr + WORD_MASK) & ~WORD_MASK);
+  }
+
+  // Copy misaligned head.
+  while (dest_addr < head) {
+    *((volatile uint8_t *)dest_addr) = *((volatile uint8_t *)src_addr);
```

Comment:
Why is this volatile? The only reason you want this is to stop the compiler from deleting the operations, and that's not going to happen.